File: Algorithm Learning/Sort Algorithm/MergeSort.java
/**
 * Created by fred on 2018/6/7.
 */
public class MergeSort {
    private static Comparable[] temp;//用作merge

    public static void main(String[] args){
        Integer[] arr = {5, 2, 76, 32, 14, 67};
        sort(arr);
        for(int k = 0; k < arr.length; k++){
            System.out.println(arr[k]);
        }
    }
    public static void sort(Comparable[] a){
        temp = new Comparable[a.length];
        sort(a, 0, a.length - 1);
    }

    public static void sort(Comparable[] a, int low, int high){
        if(high <= low){
            return;
        }
        int mid = (low + high) / 2;
        sort(a, low, mid);
        sort(a, mid + 1, high);
        merge(a, low, mid, high);
    }

    private static void merge(Comparable[] a, int low, int mid, int high) {
        int i = low, j = mid + 1;
        for(int k = low; k <= high; k++){
            temp[k] = a[k];
        }
        for(int k = low; k <= high; k++){
            if(i > mid){
                a[k] = temp[j++];
            }else if(j > high){
                a[k] = temp[i++];
            }else if(a[j].compareTo(a[i]) < 0){
                a[k] = temp[j++];
            }else {
                a[k] = temp[i++];
            }
        }
    }
}

File: Algorithm Learning/Sort Algorithm/merge_sort.py
def merge_sort(list):
    if len(list) < 2:
        return list
    else:
        mid = len(list) // 2
        lefthalf = list[:mid]
        righthalf = list[mid:]

        merge_sort(lefthalf)
        merge_sort(righthalf)

        i = 0
        j = 0
        k = 0
        while i < len(lefthalf) and j < len(righthalf):
            if lefthalf[i] < righthalf[j]:
                list[k] = lefthalf[i]
                i += 1
                k += 1
            else:
                list[k] = righthalf[j]
                j += 1
                k += 1
        while i < len(lefthalf):
            list[k] = lefthalf[i]
            i += 1
            k += 1
        while j < len(righthalf):
            list[k] = righthalf[j]
            j += 1
            k ++ 1

if __name__ == '__main__':
    list = [54,26,93,17,77,31,44,55,20]
    merge_sort(list)
    print(list)

